src/source-updater.js

```diff
@@ -13,128 +173,172 @@ import noop from './util/noop';
-    this.timestampOffset_ = 0;
+export default class SourceUpdater extends videojs.EventTarget {
+  constructor(mediaSource) {
+    super();
     this.mediaSource = mediaSource;
-    this.processedAppend_ = false;
-    this.type_ = type;
-    this.mimeType_ = mimeType;
-    this.logger_ = logger(`SourceUpdater[${type}][${mimeType}]`);
+    this.logger_ = logger('SourceUpdater');
+    // initial timestamp offset is 0
+    this.audioTimestampOffset_ = 0;
+    this.videoTimestampOffset_ = 0;
+    this.queue = [];
+    this.queuePending = {
+      audio: null,
+      video: null
+    };
+  }
 
-    if (mediaSource.readyState === 'closed') {
-      mediaSource.addEventListener(
-        'sourceopen', this.createSourceBuffer_.bind(this, mimeType, sourceBufferEmitter));
-    } else {
-      this.createSourceBuffer_(mimeType, sourceBufferEmitter);
-    }
+  ready() {
+    return !!(this.audioBuffer || this.videoBuffer);
   }
 
-  createSourceBuffer_(mimeType, sourceBufferEmitter) {
-    this.sourceBuffer_ = this.mediaSource.addSourceBuffer(mimeType);
+  createSourceBuffers(codecs) {
+    if (this.ready()) {
+      // already created them before
+      return;
+    }
 
-    this.logger_('created SourceBuffer');
+    if (this.mediaSource.readyState === 'closed') {
+      this.sourceopenListener_ = this.createSourceBuffers.bind(this, codecs);
+      this.mediaSource.addEventListener('sourceopen', this.sourceopenListener_);
+      return;
+    }
 
-    if (sourceBufferEmitter) {
-      sourceBufferEmitter.trigger('sourcebufferadded');
+    if (codecs.audio) {
+      this.audioBuffer = this.mediaSource.addSourceBuffer(
+        `audio/mp4;codecs="${codecs.audio}"`);
+      this.audioBuffer.removing = false;
+      this.logger_(`created SourceBuffer audio/mp4;codecs="${codecs.audio}`);
+    }
 
-      if (this.mediaSource.sourceBuffers.length < 2) {
-        // There's another source buffer we must wait for before we can start updating
-        // our own (or else we can get into a bad state, i.e., appending video/audio data
-        // before the other video/audio source buffer is available and leading to a video
-        // or audio only buffer).
-        sourceBufferEmitter.on('sourcebufferadded', () => {
-          this.start_();
-        });
-        return;
-      }
+    if (codecs.video) {
+      this.videoBuffer = this.mediaSource.addSourceBuffer(
+        `video/mp4;codecs="${codecs.video}"`);
+      this.videoBuffer.removing = false;
+      this.logger_(`created SourceBuffer video/mp4;codecs="${codecs.video}"`);
     }
 
+    this.trigger('ready');
     this.start_();
   }
 
   start_() {
     this.started_ = true;
 
-    // run completion handlers and process callbacks as updateend
-    // events fire
-    this.onUpdateendCallback_ = () => {
-      let pendingCallback = this.pendingCallback_;
-
-      this.pendingCallback_ = null;
-      this.sourceBuffer_.removing = false;
-
-      this.logger_(`buffered [${printableRange(this.buffered())}]`);
```

Comment:
this logic is now in the `onUpdateend` function right above the class for this file.